engine/api/build_log.go

```diff
@@ -19,33 +20,107 @@ import (
 	"github.com/ovh/cds/sdk"
 )
 
-func getBuildLogsHandler(w http.ResponseWriter, r *http.Request, db *gorp.DbMap, c *context.Ctx) error {
-
+func getStepBuildLogsHandler(w http.ResponseWriter, r *http.Request, db *gorp.DbMap, c *context.Ctx) error {
 	// Get pipeline and action name in URL
 	vars := mux.Vars(r)
 	projectKey := vars["key"]
 	pipelineName := vars["permPipelineKey"]
 	buildNumberS := vars["build"]
 	appName := vars["permApplicationName"]
+	pipelineActionIDString := vars["actionID"]
+	stepOrderString := vars["stepOrder"]
 
-	// Get offset
-	err := r.ParseForm()
+	stepOrder, errInt := strconv.Atoi(stepOrderString)
+	if errInt != nil {
+		return sdk.ErrInvalidID
+	}
+
+	pipelineActionID, errPA := strconv.ParseInt(pipelineActionIDString, 10, 64)
+	if errPA != nil {
+		return sdk.ErrInvalidID
+	}
+
+	var env *sdk.Environment
+	envName := r.FormValue("envName")
+	if envName == "" || envName == sdk.DefaultEnv.Name {
+		env = &sdk.DefaultEnv
+	} else {
+		var errEnv error
+		env, errEnv = environment.LoadEnvironmentByName(db, projectKey, envName)
+		if errEnv != nil {
+			log.Warning("getStepBuildLogsHandler> Cannot load environment %s: %s\n", envName, errEnv)
+			return sdk.ErrUnknownEnv
+		}
+
+	}
+
+	if env.ID != sdk.DefaultEnv.ID && !permission.AccessToEnvironment(env.ID, c.User, permission.PermissionRead) {
+		log.Warning("getStepBuildLogsHandler> No enought right on this environment %s: \n", envName)
+		return sdk.ErrForbidden
+	}
+
+	// Check that pipeline exists
+	p, err := pipeline.LoadPipeline(db, projectKey, pipelineName, false)
 	if err != nil {
-		log.Warning("getBuildLogsHandler> cannot parse form: %s\n", err)
-		return err
+		log.Warning("getStepBuildLogsHandler> Cannot load pipeline %s: %s\n", pipelineName, err)
+		return sdk.ErrPipelineNotFound
+	}
 
+	// Check that application exists
+	a, err := application.LoadApplicationByName(db, projectKey, appName)
+	if err != nil {
+		log.Warning("getStepBuildLogsHandler> Cannot load application %s: %s\n", appName, err)
+		return sdk.ErrApplicationNotFound
 	}
-	offsetS := r.FormValue("offset")
-	var offset int64
-	if offsetS != "" {
-		offset, err = strconv.ParseInt(offsetS, 10, 64)
+
+	// if buildNumber is 'last' fetch last build number
+	var buildNumber int64
+	if buildNumberS == "last" {
+		bn, err := pipeline.GetLastBuildNumberInTx(db, p.ID, a.ID, env.ID)
```

Comment:
nop, start a `tx` in the handler; use `GetLastBuildNumberIn(...)` and remove the function `GetLastBuildNumberInTx`